internal/app/machined/pkg/controllers/time/sync.go

```diff
@@ -0,0 +1,178 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package time
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/AlekSi/pointer"
+	"github.com/talos-systems/os-runtime/pkg/controller"
+	"github.com/talos-systems/os-runtime/pkg/resource"
+	"github.com/talos-systems/os-runtime/pkg/state"
+
+	v1alpha1runtime "github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
+	"github.com/talos-systems/talos/internal/pkg/ntp"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/resources/config"
+	"github.com/talos-systems/talos/pkg/resources/time"
+	"github.com/talos-systems/talos/pkg/resources/v1alpha1"
+)
+
+// SyncController manages v1alpha1.TimeSync based on configuration and NTP sync process.
+type SyncController struct {
+	V1Alpha1Mode v1alpha1runtime.Mode
+	NewNTPSyncer NewNTPSyncerFunc
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *SyncController) Name() string {
+	return "time.SyncController"
+}
+
+// ManagedResources implements controller.Controller interface.
+func (ctrl *SyncController) ManagedResources() (resource.Namespace, resource.Type) {
+	return v1alpha1.NamespaceName, time.StatusType
+}
+
+// NTPSyncer interface is implemented by ntp.Syncer, interface for mocking.
+type NTPSyncer interface {
+	Run(ctx context.Context)
+	Synced() <-chan struct{}
+	EpochChange() <-chan struct{}
+	SetTimeServers([]string)
+}
+
+// NewNTPSyncerFunc function allows to replace ntp.Syncer with the mock.
+type NewNTPSyncerFunc func(*log.Logger, []string) NTPSyncer
+
+// Run implements controller.Controller interface.
+//
+//nolint: gocyclo,cyclop
```

Comment:
```suggestion
//nolint:gocyclo,cyclop
```

In other places too.

I will look into writing a [ruleguard](https://github.com/quasilyte/go-ruleguard) rule for that.

Reply:
fixed